runtime/src/bank.rs

```diff
@@ -12724,7 +12698,7 @@ pub(crate) mod tests {
                 .accounts_db
                 .accounts_index
                 .ref_count_from_storage(&key5.pubkey()),
-            expected_ref_count_for_keys_only_in_slot_2
+            expected_ref_count_for_keys_in_both_slot1_and_slot2,
```

Comment:
@brooksprumo these tests were super helpful for debugging 😄 

For this test here there was a peculiar case. I think the correct ref count here should be `1`, since `pubkey5` was added in both `bank1` and `bank1` and then we dropped/purged `bank1`. The old test expected `expected_ref_count_for_keys_only_in_slot_2 == 2` references, which seems wrong in retrospect. With this change, it seems to be correctly reporting a refcount of 1.

Reply:
Glad they helped!

Yeah, I was confused about the ref count of 2 as well. I took a step back in this particular test case to figure out what I _really_ wanted to test out and avoid testing other components (if possible). That's why I compared key5 here to key2, because I knew that if key2 had the right value, then key5 should have that same value, whatever it actually is.

I think there is/was another issue that caused the ref count to be 2 instead of 1, which looks like you've found and fixed (yay!). To make sure I understand right, shouldn't the original logic still work in the assert too? Specifically:

```
let expected_ref_count_for_keys_in_both_slot1_and_slot2 = bank2
    .rc
    .accounts
    .accounts_db
    .accounts_index
    .ref_count_from_storage(&key2.pubkey());
```

Reply:
@brooksprumo, you'll be surprised to find that the original logic you pointed out returns `2` instead of `1` for `expected_ref_count_for_keys_in_both_slot1_and_slot2`. This is because `key2` is touched in both `bank0` and `bank1`.

Now you may that that because the only *explicit* transaction we did in `bank0` was a transfer to `key2`, then when that update is outdated by a transfer on a later *rooted* bank `bank2.transfer(4, &mint_keypair, &key2.pubkey()).unwrap();`, then `clean_accounts()` should mark slot `0` as dead and unref `key2`. However, `bank0` actually has a lot of additional keys initialized from genesis that have not been outdated yet, namely sysvars and various program accounts, here's a list!

```
updating slot 0, pubkey 3ywG1DkxsBYvS6oq4cmeCyzPJzDYeX2eLhEwyJjJaocj
updating slot 0, pubkey SysvarFees111111111111111111111111111111111
updating slot 0, pubkey 11111111111111111111111111111111
updating slot 0, pubkey Vote111111111111111111111111111111111111111
updating slot 0, pubkey Stake11111111111111111111111111111111111111
updating slot 0, pubkey Config1111111111111111111111111111111111111
updating slot 0, pubkey So11111111111111111111111111111111111111112
updating slot 0, pubkey SysvarStakeHistory1111111111111111111111111
updating slot 0, pubkey SysvarC1ock11111111111111111111111111111111
updating slot 0, pubkey SysvarRent111111111111111111111111111111111
updating slot 0, pubkey SysvarEpochSchedu1e111111111111111111111111
updating slot 0, pubkey SysvarRecentB1ockHashes11111111111111111111
updating slot 0, pubkey 3ywG1DkxsBYvS6oq4cmeCyzPJzDYeX2eLhEwyJjJaocj
updating slot 0, pubkey DMfWhsGGtwa4otGUrAx1fNMRbSGZ3pXdxtuw67k2G6wr
```

Updated in later slots:

```
updating slot 1, pubkey SysvarS1otHashes111111111111111111111111111
updating slot 1, pubkey SysvarC1ock11111111111111111111111111111111
updating slot 1, pubkey SysvarFees111111111111111111111111111111111
updating slot 1, pubkey 3ywG1DkxsBYvS6oq4cmeCyzPJzDYeX2eLhEwyJjJaocj
updating slot 1, pubkey 9dG5h4dsfz3vCy9C8c4Fhod4F96YtJcHe9okwht8ouit
updating slot 1, pubkey 36nj3aidYASj5kndMV2GsjcGB8egiEzGLiBAZFXcGhs8
updating slot 1, pubkey Hu8UuzSyQStcsfMFsxNepzG7PN2yKDvH8Qez32WDR5D8
updating slot 2, pubkey SysvarS1otHashes111111111111111111111111111
updating slot 2, pubkey SysvarC1ock11111111111111111111111111111111
updating slot 2, pubkey SysvarFees111111111111111111111111111111111
updating slot 2, pubkey 3ywG1DkxsBYvS6oq4cmeCyzPJzDYeX2eLhEwyJjJaocj
updating slot 2, pubkey DMfWhsGGtwa4otGUrAx1fNMRbSGZ3pXdxtuw67k2G6wr
updating slot 2, pubkey 3ywG1DkxsBYvS6oq4cmeCyzPJzDYeX2eLhEwyJjJaocj
updating slot 2, pubkey 44y4dbRFM2Pijb6nRFJjbTaU96ut62tbGhiFnFjGY5Zs
updating slot 2, pubkey Hu8UuzSyQStcsfMFsxNepzG7PN2yKDvH8Qez32WDR5D8
updating slot 2, pubkey SysvarS1otHistory11111111111111111111111111
```